### ISSUE TITLE
Should the handle be an object?

### ISSUE THREAD

**Issue by eladalon1983** [w3c/mediacapture-handle#68]
At the moment, the [handle](https://w3c.github.io/mediacapture-handle/identity/index.html#dom-capturehandle-handle) field is a string. This limits it to stringable objects rather than serializable objects. If we change to an object, applications will be able to expose [CropTargets](https://w3c.github.io/mediacapture-region/#dom-croptarget) as well.

```js
navigator.mediaDevices.setCaptureHandleConfig({
  exposeOrigin: true,
  permittedOrigins: ['*'],
  handle: {
    field1: "value1",
    field2: "value2",
    cropTargets: {
      label1: cropTarget1,
      label2: cropTarget2,
    },
  },
});
```

@youennf, I think you were amenable to this during TPAC 2022...?
@jan-ivar, your opinion?

Note that I still intend to pursue orthogonal proposals for:
1. **Structured** exposure of CropTargets.
2. **Structured** exposure of a suggested contentHint.
3. Exposing a MessagePort (transferable but not serializeable, btw) for bidirectional **unstructured** messaging. (The `handle` is one way and immediate, which is important [a] for Conditional Focus without exceeding the task-end limit, and [b] for capturers that don't wish to communicate with the capturee and alert it to the existence of a capture session.)

**Comment by youennf** [comment]
> @youennf, I think you were amenable to this during TPAC 2022...?

I think this is a reasonable answer to the crop target on another tab use case.
An alternative to consider is to enable MessagePort level communication.

One thing to consider is that the current 1024 bytes size might be harder to define with objects.

As a side note, and this also applies to structured exposure of CropTargets, if a web application starts to create CropTargets proactively, UAs should make the creation of such objects cheap and only start expensive operations when these objects are actually used to crop. This is another argument in favour of creating CropTarget synchronously.

**Comment by eladalon1983** [comment]
> One thing to consider is that the current 1024 bytes size might be harder to define with objects.

I intended that limitation to defang potential attacks where a captured site overwhelms a capturing site with extremely large messages. That's probably¹ not terribly concerning, though, as the captured site would have to pay as much for the attack as would the victim, and the cost would be paid without the attacker even knowing if there is a victim, or who it is. I think it's safe to drop this limitation².

> An alternative to consider is to enable MessagePort level communication.

There is value in the capturer being able to read something about the capturee, without letting the capturee know that there is a capturer.
* It prevents attacks like the one discussed in the previous paragraph.
* It simplifies things for some Web developers. For capurees that only have on static thing to report, it's easier to set the CaptureHandleConfig once, than to register a handler that responds to queries over a MessagePort.

> As a side note, and this also applies to structured exposure of CropTargets, if a web application starts to create CropTargets proactively, UAs should make the creation of such objects cheap and only start expensive operations when these objects are actually used to crop. This is another argument in favour of creating CropTarget synchronously.

Agreed. (But there are other arguments both ways.)

For now, I am not concerned, because I expect a very limited number of CropTargets. But, yes, that is indeed a valid argument. If we later observe that sites often expose several CropTargets, this argument could increase in weight.

--
[1] Or wdyt?
[2] I intend to ask Chrome's security experts.

**Comment by eladalon1983** [comment]
To help drive collaboration between loosely-coupled apps, I've presented the following earlier today.

Assume:

```webidl
dictionary ContentHints {
  DOMString audio;
  DOMString video;
};

dictionary CropTargetData {
  DOMString name;
  CropTarget cropTarget;
  ContentHints suggestedContentHint;
};

dictionary CaptureHandleConfig {
  ...  // Pre-existing.

  // Suggested Content-Hint for uncropped video.
  ContentHints suggestedContentHints;

  // CropTargets of interest.
  sequence<CropTargetData> cropTargets;
};
```

Then a VC app could do something like this:
```js
const stream = navigator.mediaDevices.getDisplayMedia(constraints);
const [track] = stream.getVideoTracks();
const captureHandle = track.getCaptureHandle();

if (Object.keys(captureHandle?.cropTargets).length > 1) {
  // 1. Cycle through the crop-targets,
  // grabbing one frame with each.
  const thumbnails = GetThumbnails(track);

  // 2. Display an in-content picker.
  // The user chooses if they want to share
  // the entire tab, or only a specific part of it.
  const target = PromptUser(track, thumbnails);

  // 3. Apply user’s choice.
  track.cropTo(target);  // No-op if `target` is `undefined`.

  // 4. Transmit remotely.
  TransmitRemotely(track);
}
```

To put a cherry on top, in the `TransmitRemotely()` step, the VC app can - if it chooses - use the suggestedContentHint tied to the specific region.


This means that Teams, Zoom and Meet could all offer great **functionality** and **performonace** when capturing YouTube, Hulu, Slides, PowerPoint or anything else.

**Comment by jan-ivar** [comment]
TL;DR: I don't think `handle` should be an object, as this seems like a design break.

captureHandle.[handle](https://w3c.github.io/mediacapture-handle/identity/index.html#dom-capturehandle-handle) was billed as an _"identity"_ string for _"[Bootstrapping](https://w3c.github.io/mediacapture-handle/identity/)"_ _"[a connection](https://w3c.github.io/mediacapture-handle/identity/index.html#generic-problem-desc)"_ [[1]](https://w3c.github.io/mediacapture-handle/identity/index.html#use-case-1):

> The means for transmitting these messages are outside the scope of this document. Some options are:
> 
> * Shared cloud infrastructure.
> * Messaging via a worker. (Note: Storage Partitioning might disrupt this option.)
> * A rudimentary messaging API might be added expressly for this purpose.

Letting `handle` accept serializable objects doesn't fit with that design, and takes us instead another step closer to reinventing postMessage, which I warned about in https://github.com/w3c/mediacapture-handle/issues/11.

If storage partitioning has made out-of-band messaging nonviable, CropTargets aren't serializable, and we're trying to send objects through` handle`, then we seem overdue for the third option IMHO... a rudimentary messaging API.

A way forward on that might be:
1. Address https://github.com/w3c/mediacapture-handle/issues/11 first
2. Move captureHandle to captureController first to avoid message fan-out so we don't spam every track
3. Specify a port on each captureHandle and a specialized port somewhere on the capturee side that caches the last message sent on it ahead of capture, to surface it on the captureController upon gDM success.

If the capturer never responds, then the capturee never learns it's being captured.

**Comment by eladalon1983** [comment]
> If the capturer never responds, then the capturee never learns it's being captured.

I object to many parts of the previous message, but I think we can zero in on one point first - the one I quoted. We fundamentally disagree on the alternative of MessagePort and what it would look like. I don't think the model you suggest is even possible, let alone desirable.

First, feasibility. A tab can be captured by multiple capturers. A design centered around a MessagePort will require that the captured page register an event handler to receive a MessagePort after the capturer "dials in" by calling `CaptureController.getMessagePort()`. That means that captured pages cannot send any messages ahead of knowing if they're being captured.

Second, even if your model were possible, it's evident that you noticed the following issue and tried to sidestep it - it's highly undesirable to require that captured pages routinely send the same message, which is why you introduced the caching mechanism. That is, you **reinvented the handle**. I now rest my case.

**Comment by jan-ivar** [comment]
Let's take a step back then. My main point was:
> we seem overdue for the third option IMHO... a rudimentary messaging API.

Do we agree on this?

This issue already tries to reinvent handle, opening up for alternative ideas. Neither a string identifier nor a traditional message port seem to satisfy requirements, so some innovation seems unavoidable. Since iteration has failed us, I think we need to reexamine the requirements, and figure out what shape makes most sense for them.

With this issue and https://github.com/w3c/mediacapture-handle/issues/11, `handle` is veering toward reinventing [broadcastChannel.postMessage](https://html.spec.whatwg.org/multipage/web-messaging.html#dom-broadcastchannel-postmessage-dev) except across origins IMHO. At the same time, [permittedOrigins](https://w3c.github.io/mediacapture-handle/identity/index.html#dom-capturehandleconfig-permittedorigins) smells like a list of [targetOrigin](https://html.spec.whatwg.org/multipage/window-object.html#dom-windowpostmessageoptions-targetorigin)s, a feature [window.postMessage](https://html.spec.whatwg.org/multipage/web-messaging.html#posting-messages). And lastly, a [MessagePort](https://html.spec.whatwg.org/multipage/web-messaging.html#messageport) can only be transferred over the latter (since broadcast naturally cannot transfer).

The previous paragraph mentions every object that implements `postMessage()` AFAIK, and while none of them fit, the idea that `postMessage` might at least be a shape we should consider hopefully seems reasonable. 

So for sake of illustration only, we might imagine a broadcast-channel-like API with targetOrigin, something like:
```js
// capturee
navigator.mediaDevices.captureConfig.postMessage({id, hint, cropTarget}, "*");
navigator.mediaDevices.captureConfig.onmessage = e => iRecognize(e.origin) && my1on1Port = e.port;

// capturer
captureController.onmessage = e => {
  if (iRecognize(e.origin)) {
    const {port1, port2} = new MessageChannel();
    my1on1Port = port1;
    captureController.postMessage(port2, targetOrigin: e.origin);
  }
```

**Comment by eladalon1983** [comment]
First, this illustration assumes that the capturee can only be captured by a single capturer. That is false, as I have pointed out in my last message. I have also explained that this should not be abstracted away for the sake of illustration, because it has corollaries. Namely, the only model that **will** work for producing a MessagePort, is one in which the capturee registers to be notified when the capturer tries to establish a MessagePort to it, and this notification has ramifications on issues which you have expressed concern for - self-censorship.

Second, this **does not work** for loosely-coupled applications, which would not have advanced agreement on the protocol for the messages sent over the MessagePort. Again - a use-case you have previously said you cared about.

**Comment by jan-ivar** [comment]
No, the illustration is of a new API where `mediaDevices.captureConfig.postMessage()` is broadcast (one-to-many) and cannot accept transferable objects, whereas the other direction `captureController.postMessage` can (since there is only one recipient).

Self-censorhip is already possible in the current spec once the capturer establishes a "connection".

I don't think we should solve more features for loosely-coupled applications atm.

**Comment by eladalon1983** [comment]
> Self-censorhip is already possible in the current spec once the capturer establishes a "connection".

The current spec does NOT force an establishment of a connection. My [proposal](https://github.com/w3c/mediacapture-handle/issues/68#issuecomment-1282898580) shows a simple mechanism for publishing actionable information without requiring a connection. That proposal was designed **precisely** to address the concerns you have previously raised, which were (1) self-censorship and (2) usefulness to loosely-coupled apps.

> I don't think we should solve more features for loosely-coupled applications atm.

Could you please explain this 180-degrees change in your position?

**Comment by jan-ivar** [comment]
First, thank you for attempting to address concerns you thought I might have. I appreciate that, but it's not that simple.

It goes back to use cases. I found users having [nextslide](https://w3c.github.io/mediasession/#dom-mediasessionaction-nextslide) and [previousslide](https://w3c.github.io/mediasession/#dom-mediasessionaction-previousslide) on a loosely-coupled captured page compelling. But I'm fairly confident those may also end up becoming keyboard hardware hotkeys someday, which is a high bar.

But those also opened up a long tail of ideas that didn't make it ("firstslide", "lastslide", https://github.com/w3c/mediasession/issues/282) and maybe never should, because there's a risk and a cost to baking into the web platform forever what we think apps want today.

I don't find loosely-coupled pages advertising crop regions compelling enough to bake into the web platform now.
I don't find loosely-coupled pages advertising content hints compelling enough to bake into the web platform now.

It's not clear to me we're ready to say how this should work forever. App-controls also encroach on what's a fully user-controlled feature today, so there are risks they may sometimes undermine user choice if we're not careful.

Since this spec already enables apps to experiment in this space, I'd rather wait an see how they solve this.

> > we seem overdue for the third option IMHO... a rudimentary messaging API.
> 
> Do we agree on this?

I remain curious about an answer to this question.

**Comment by jan-ivar** [comment]
I mean adding a MessagePort. It seems to me we can make progress there, which should give collaborating apps everything they need to experiment with handling these use cases.

**Comment by eladalon1983** [comment]
> > I don't think we should solve more features for loosely-coupled applications atm.
>
> Could you please explain this 180-degrees change in your position?

It occurs to me that #69 remove the need for Capture Handle Identity altogether, but only works for tightly-coupled apps.

**Comment by jan-ivar** [comment]
That is a misunderstanding.

**Comment by jan-ivar** [comment]
I sense generally good agreement on the use case here, with the only differences being over API shape. This seems like something we should be able to make progress on.

My concerns that I don't believe have been addressed here are:
1. Calling [setCaptureHandleConfig](https://w3c.github.io/mediacapture-handle/identity/index.html#dom-mediadevices-setcapturehandleconfig) rapidly doesn't seem web compatible wrt what a capturer sees. From its algorithm, it looks like messages can get missed due to timing. postMessage solves this using a queue. 
2. There's no surfacing of [deserialization errors](https://github.com/w3c/webcodecs/issues/635#issuecomment-1412798059), which seems necessary to handle arbitrary objects (e.g. [FileReaderSync](https://www.w3.org/TR/FileAPI/#FileReaderSync) and several other objects cannot be instantiated in Window). postMessage has [onmessageerror](https://html.spec.whatwg.org/#handler-window-onmessageerror) for this.

While (1) might be addressed by adding the message to the event, this would seem to go against [§ 7.7. Use plain Events for state](https://w3ctag.github.io/design-principles/#state-and-subclassing), perhaps suggesting that recreating (post) message-passing behavior using events is not a desirable pattern.

**Comment by eladalon1983** [comment]
> I sense generally good agreement on the use case here, with the only differences being over API shape.

I think there are multiple use cases here, and we only have agreement on some. It may be that the use-cases where we disagree are appear uninteresting to you, but they're important to me, so I disagree with the characterization of "general agreement".

To clarify, we disagree on:

* Communicating potential CropTargets to an unknown capturer.
* Communicating suggested [content hints](https://www.w3.org/TR/mst-content-hint/) to an unknown capturer.

I intend to get back to this topic (future reshuffles of duties and priorities notwithstanding). When that happens, if you still find these uninteresting, to the point of objecting to tackling the problem, then I don't see a way forward for me other than to fork the document and proceed with other interested parties. If you can think of a nicer way forward, please do share it.

**Comment by eladalon1983** [comment]
> From its algorithm, it looks like messages can get missed due to timing.

How are [the events](https://w3c.github.io/mediacapture-handle/identity/index.html#events) insufficient for ensuring that nothing is missed?


> There's no surfacing of deserialization errors,

* Presently, no serialization errors can occur, because we only accepts strings.
* If we extend along the lines of [this approach](https://github.com/w3c/mediacapture-handle/issues/68#issuecomment-1282898580), no serialization errors are possible either.
* If we change the [handle](https://w3c.github.io/mediacapture-handle/identity/index.html#dom-capturehandleconfig-handle) to be `object` or `(DOMString or object)`, we can specify that [setCaptureHandleConfig()](https://w3c.github.io/mediacapture-handle/identity/index.html#set-capture-handle-config) can raise a synchronous error if it cannot guarantee that the object is serializable. That seems simple enough, does it not?

> this would seem to go against § 7.7. Use plain Events for state,

As mentioned, I don't see any problems with the current approach; in fact, you have yourself convinced me to stop exposing the message on the event, because we both agreed it was unnecessary - see [here](https://github.com/w3c/mediacapture-handle/issues/50).

But assume for the sake of argument that adding the message back were necessary. The first words in `§ 7.7. Use plain Events for state` are: **"Where possible"**.
